kratos/utilities/constraint.h

```diff
@@ -0,0 +1,294 @@
+    typedef std::vector<std::size_t> EquationIdVectorType;
+
+    typedef typename TDenseSpace::MatrixType LocalSystemMatrixType;
+    typedef typename TDenseSpace::VectorType LocalSystemVectorType;
+    typedef typename TSparseSpace::DataType TDataType;
+    typedef typename TSparseSpace::MatrixType TSystemMatrixType;
+    typedef typename TSparseSpace::VectorType TSystemVectorType;
+
+    ///@name Life Cycle
+    ///@{
+
+    /**
+	* Creates a Constraint object
+	*/
+    Constraint(std::string iName = "default", bool iIsActive = true) : mActive(iIsActive), mName(iName)
+    {
+    }
+    /// Destructor.
+    virtual ~Constraint(){};
+
+    ///@}
+
+    ///@name Access
+    ///@{
+
+    /**
+	* Clears the maps contents
+	*/
+    void Clear()
+    {
+        mConstraintEquationContainer.Clear();
+    }
+
+    /**
+	* Get the MasterDOFs vector for this slave
+	* @return MasterDOFs vector for this slave
+	*/
+    virtual const ConstraintEquation &GetConstraintEquation (DofType &SlaveDof)
+    {
+        return mConstraintEquationContainer.GetConstraintEquation(SlaveDof);
+    }
+
+    /**
+    * Adds a constraints between the given slave and master with a weight. 		
+	*/
+    // Takes in a slave dof equationId and a master dof equationId
+    virtual void AddConstraint(DofType &SlaveDof, DofType &MasterDof, double weight, double constant = 0.0)
+    {
+        mConstraintEquationContainer.AddConstraint(SlaveDof, MasterDof, weight, constant);
+    }
+
+    /**
+     *  Does necessary operations to setup the constraint.
+     */
+    virtual void SetUp(NodesContainerType &Nodes)
+    {
+    }
+
+    /**
+     *  Does necessary operations on the constraint before the build of master stiffness matrix is commenced
+     */
+    virtual void ExecuteBeforeBuilding(NodesContainerType &Nodes)
+    {
+    }
+    /**
+     *  Does necessary operations on the constraint after the build of master stiffness matrix is commenced
+     */
+    virtual void ExecuteAfterBuilding(NodesContainerType &Nodes)
+    {
+    }
+
+    /**
+     *  Does necessary operations on global symtem Ax=b on the constraint before they are solved.
+     */    
+    virtual void ExecuteBeforeSolving(TSystemMatrixType &A,
+                                      TSystemVectorType &Dx,
+                                      TSystemVectorType &b)
+    {
+    }
+
+    /**
+     *  Does necessary operations on global symtem Ax=b on the constraint after they are solved.
+     */    
+    virtual void ExecuteAfterSolving(TSystemMatrixType &A,
+                                     TSystemVectorType &Dx,
+                                     TSystemVectorType &b)
+    {
+    }
+
+    /**
+     *  Does necessary operations on the element freedom table so as to construct the global stiffness matrix.
+     *  Mainly to build the sparsity pattern for the global stiffness matrix.
+     */        
+    virtual void Element_ModifyEquationIdsForConstraints(Element &rCurrentElement,

```

Comment:
Instead of Element_ and Condition_ you could overload the operations ModifyEquationIds and Apply Constraints.

Reply:
Good point. I will do that.

Reply:
I also missed this, but there is a typo in the function GetNumbeOfMasterDofsForSlave